### PR TITLE
Fix/refresh token

### PR DESCRIPTION
**Pull request by erzohrakhan**
Refresh token API added

### CODE REVIEW

File: iridium-core-server/src/main/java/software/iridium/api/controller/AuthorizationController.java [hunk 1]
```diff
@@ -70,4 +70,10 @@ public AccessTokenResponse exchange(
     logger.info("finalizing authorization");
     return authorizationService.exchange(servletRequest, params);
   }
+
+  @PostMapping(value = "/oauth/token/refresh")
+  public AccessTokenResponse refreshToken(@RequestParam Map<String, String> params) {
+    logger.info("access token request through refresh token");
+    return authorizationService.refreshToken(params);
+  }
```
Comment (joshfischer1108) [inline iridium-core-server/src/main/java/software/iridium/api/controller/AuthorizationController.java:78]: Let's add a test for this method.

File: iridium-core-server/src/main/java/software/iridium/api/service/AuthorizationService.java [hunk 2]
```diff
@@ -380,10 +380,6 @@ public AccessTokenResponse exchange(
             accessTokenRepository.save(
                 accessTokenInstantiator.instantiate(authorizationCode.getIdentityId()));
 
-        final var refreshToken = refreshTokenInstantiator.instantiate(accessToken.getAccessToken());
-        accessToken.setRefreshToken(refreshToken);
-        refreshToken.setAccessToken(accessToken);
```
Comment (joshfischer1108) [inline iridium-core-server/src/main/java/software/iridium/api/service/AuthorizationService.java:385]: We need to keep the association from the `refreshToken` to the `accessToken`.  

e.g. 
` refreshToken.setAccessToken(accessToken);`

File: iridium-core-server/src/main/java/software/iridium/api/service/AuthorizationService.java [hunk 3]
```diff
@@ -445,4 +441,50 @@ public String proxyAuthorizationRequestToProvider(
...
+    checkArgument(
+        attributeValidator.isNotBlank(
+            params.getOrDefault(AuthorizationCodeFlowConstants.REFRESH_TOKEN.getValue(), "")),
+        "Refresh_token must not be blank");
+
+    checkArgument(
+        attributeValidator.equals(
+            params.get(AuthorizationCodeFlowConstants.GRANT_TYPE.getValue()),
+            AuthorizationCodeFlowConstants.REFRESH_TOKEN_GRANT_TYPE.getValue()),
+        "Grant_type value must be set to \"refresh_token\"");
```
Comment (joshfischer1108) [inline iridium-core-server/src/main/java/software/iridium/api/service/AuthorizationService.java:467]: These `checkArgument`'s should be in a validator class.  You can see examples of them [here](https://github.com/IridiumIdentity/iridium/tree/main/iridium-core-server/src/main/java/software/iridium/api/validator)

File: iridium-core-server/src/test/java/software/iridium/api/instantiator/AccessTokenEntityInstantiatorTest.java [hunk 4]
```diff
@@ -48,10 +52,13 @@ public void instantiate_AllGood_BehavesAsExpected() {
...
         .thenReturn(accessTokenValue);
+    when(mockRefreshTokenEntityInstantiator.instantiate(accessTokenValue))
+        .thenReturn(mockRefreshTokenEntity);
 
     final var response = subject.instantiate(identityId);
 
     verify(mockTokenGenerator).generateAccessToken(same(identityId), any(Date.class));
+    verify(mockRefreshTokenEntityInstantiator).instantiate(same(accessTokenValue));
 
     assertThat(response.getAccessToken(), is(equalTo(accessTokenValue)));
```
Comment (joshfischer1108) [inline iridium-core-server/src/test/java/software/iridium/api/instantiator/AccessTokenEntityInstantiatorTest.java:63]: Let's also assert that the refreshToken is properly associated with accessToken. 

e.g. 
```
    assertThat(response.getRefreshToken(), is(sameInstance(mockRefreshTokenEntity)));
```

File: iridium-core-server/src/test/java/software/iridium/api/mapper/AccessTokenResponseMapperTest.java [hunk 5]
```diff
@@ -33,14 +34,18 @@ public void setUpForEachTestCase() {
...
     entity.setAccessToken(accessToken);
     entity.setTokenType(tokenType);
-
+    entity.setRefreshToken(refreshTokenEntity);
     final var response = subject.map(entity);
 
     assertThat(response.getAccessToken(), is(equalTo(accessToken)));
     assertNull(response.getRedirectUrl());
     assertThat(response.getTokenType(), is(equalTo(tokenType)));
+    assertThat(response.getRefreshToken(), is(equalTo(refreshToken)));
```
Comment (joshfischer1108) [inline iridium-core-server/src/test/java/software/iridium/api/mapper/AccessTokenResponseMapperTest.java:49]: I like that you updated the test here.  +1

File: iridium-core-server/src/main/java/software/iridium/api/service/AuthorizationService.java [hunk 6]
```diff
@@ -445,4 +441,50 @@ public String proxyAuthorizationRequestToProvider(
...
+
+    checkArgument(
+        attributeValidator.isNotBlank(
+            params.getOrDefault(AuthorizationCodeFlowConstants.GRANT_TYPE.getValue(), "")),
+        "grant_type must not be blank");
+
+    checkArgument(
+        attributeValidator.isNotBlank(
+            params.getOrDefault(AuthorizationCodeFlowConstants.CLIENT_ID.getValue(), "")),
+        "Client ID must not be blank");
```
Comment (joshfischer1108) [inline iridium-core-server/src/main/java/software/iridium/api/service/AuthorizationService.java:456]: parameter identifiers should match the parameter name for easier debugging.

e.g. `Client ID` should be `client_id`

File: iridium-core-server/src/main/java/software/iridium/api/service/AuthorizationService.java [hunk 7]
```diff
@@ -445,4 +441,50 @@ public String proxyAuthorizationRequestToProvider(
...
+
+    checkArgument(
+        attributeValidator.isNotBlank(
+            params.getOrDefault(AuthorizationCodeFlowConstants.REFRESH_TOKEN.getValue(), "")),
+        "Refresh_token must not be blank");
```
Comment (joshfischer1108) [inline iridium-core-server/src/main/java/software/iridium/api/service/AuthorizationService.java:461]: same here.

File: iridium-core-server/src/main/java/software/iridium/api/service/AuthorizationService.java (hunk 3 above)
Comment (joshfischer1108) [inline iridium-core-server/src/main/java/software/iridium/api/service/AuthorizationService.java:467]: same here.

File: iridium-core-server/src/main/java/software/iridium/api/util/AuthorizationCodeFlowConstants.java [hunk 8]
```diff
@@ -23,8 +23,9 @@ public enum AuthorizationCodeFlowConstants {
   AUTHORIZATION_CODE("code"),
   AUTHORIZATION_CODE_GRANT_TYPE("authorization_code"),
   GRANT_TYPE("grant_type"),
-  CODE_VERIFIER("code_verifier");
-
+  CODE_VERIFIER("code_verifier"),
+  REFRESH_TOKEN_GRANT_TYPE("refresh_token"),
+  REFRESH_TOKEN("refresh_token");
```
Comment (joshfischer1108) [inline iridium-core-server/src/main/java/software/iridium/api/util/AuthorizationCodeFlowConstants.java:28]: What's the significance of having two enums with the same value?

**Review by joshfischer1108 — COMMENTED** [review]
@erzohrakhan  This is a fantastic contribution, thank you!  I left a few comments on code blocks we should address before we merge. 

Let me know if you have any questions.

File: iridium-core-server/src/main/java/software/iridium/api/controller/AuthorizationController.java (hunk 1 above)
Comment (erzohrakhan) [reply iridium-core-server/src/main/java/software/iridium/api/controller/AuthorizationController.java]: done

File: iridium-core-server/src/main/java/software/iridium/api/service/AuthorizationService.java (hunk 2 above)
Comment (erzohrakhan) [reply iridium-core-server/src/main/java/software/iridium/api/service/AuthorizationService.java:385]: The same code is written in AccessTokenInstantiator instantiate function

File: iridium-core-server/src/main/java/software/iridium/api/service/AuthorizationService.java (hunk 3 above)
Comment (erzohrakhan) [reply iridium-core-server/src/main/java/software/iridium/api/service/AuthorizationService.java]: done

File: iridium-core-server/src/test/java/software/iridium/api/instantiator/AccessTokenEntityInstantiatorTest.java (hunk 4 above)
Comment (erzohrakhan) [reply iridium-core-server/src/test/java/software/iridium/api/instantiator/AccessTokenEntityInstantiatorTest.java:65]: +1

File: iridium-core-server/src/main/java/software/iridium/api/service/AuthorizationService.java (hunk 3 above)
Comment (erzohrakhan) [reply iridium-core-server/src/main/java/software/iridium/api/service/AuthorizationService.java]: done

File: iridium-core-server/src/main/java/software/iridium/api/service/AuthorizationService.java [hunk 9]
```diff
@@ -305,7 +305,7 @@ public AccessTokenResponse exchange(
 
       if (attributeValidator.isBlank(
           params.getOrDefault(AuthorizationCodeFlowConstants.CLIENT_ID.getValue(), ""))) {
-        throw new BadRequestException("application id blank or malformed");
+        throw new BadRequestException("client id blank or malformed");
```
Comment (joshfischer1108) [inline iridium-core-server/src/main/java/software/iridium/api/service/AuthorizationService.java:306]: Thank you for fixing.  Good catch.

File: iridium-core-server/src/main/java/software/iridium/api/service/AuthorizationService.java (hunk 3 above)
Comment (joshfischer1108) [reply iridium-core-server/src/main/java/software/iridium/api/service/AuthorizationService.java]: These `checkArgument` functions are still in the `AuthorizationService` class.  They should be in a validator class.  I listed examples of them above.

File: iridium-core-server/src/main/java/software/iridium/api/controller/AuthorizationController.java (hunk 1 above)
Comment (joshfischer1108) [reply iridium-core-server/src/main/java/software/iridium/api/controller/AuthorizationController.java]: I don't see a test.  Am I missing something?

File: iridium-core-server/src/main/java/software/iridium/api/controller/AuthorizationController.java [hunk 10]
```diff
@@ -70,4 +70,13 @@ public AccessTokenResponse exchange(
     logger.info("finalizing authorization");
     return authorizationService.exchange(servletRequest, params);
   }
+
+  @PostMapping(value = "/oauth/token/refresh")
+  public AccessTokenResponse refreshToken(
+      @RequestParam(value = "grant_type") final String grantType,
```
Comment (joshfischer1108) [inline iridium-core-server/src/main/java/software/iridium/api/controller/AuthorizationController.java:76]: I like the explicit params.  Well done. +1

**Pull request merged**